src/ucp/api/ucp.h

```diff
@@ -2732,6 +2762,38 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
                                     size_t *length, unsigned flags);
 
 
+/**
+ * @ingroup UCP_COMM
+ * @brief Non-blocking stream receive operation of structured data into a
+ *        user-supplied buffer.
+ *
+ * This routine receives data that is described by the local address @a buffer,
+ * size @a count object on the endpoint @a ep. The routine is non-blocking
+ * and therefore returns immediately. The receive operation is considered
+ * complete when the message is delivered to the buffer. If the receive
+ * operation cannot be started, the routine returns an error.
+ *
+ * @param [in]     ep       UCP endpoint that is used for the receive operation.
+ * @param [in]     buffer   Pointer to the buffer to receive the data to.
+ * @param [in]     count    Number of elements to receive into @a buffer.
+ * @param [out]    length   Size of the received data in bytes. The value is
+ *                          valid only if return code is UCS_OK.
+ * @note                    The amount of data received, in bytes, is always an
+ *                          integral multiple of the @a datatype size.
+ * @param [in]     param    Operation parameters, see @ref ucp_request_param_t
+ *
+ * @return NULL                 - The receive operation was completed
+ *                                immediately.
+ * @return UCS_PTR_IS_ERR(_ptr) - The receive operation failed.
+ * @return otherwise            - Operation was scheduled for receive. A request
+ *                                handle is returned to the application in order.
+ *                                to track progress of the operation.
+ */
+ucs_status_ptr_t ucp_stream_recv_nbx(ucp_ep_h ep, void *buffer, size_t count,
+                                     size_t *length,
+                                     const ucp_request_param_t *param);
```

Comment:
should we add STREAM_WAITALL to ucp_request_param_t flags?

Reply:
I think yes
@evgeny-leksikov any ideas?

Reply:
yes, we should, see my thoughts here https://github.com/openucx/ucx/pull/4993#discussion_r407339982

Reply:
ok, added

Reply:
also please add `ucp_stream_recv_callback_t ` type to the params (as a union of recv cbs).

Reply:
done: added new callback type